---
title: "StatMethodsHW3A"
author: "Joshua Burkhart"
date: "January 26, 2016"
output: 
  pdf_document: 
    fig_width: 9
    fig_height: 6
    latex_engine: xelatex
---

# BMI 651: HW3 A

```{r global_options, echo=FALSE, include=FALSE, error=FALSE}
knitr::opts_chunk$set(fig.path = "Figs/",
                      message = FALSE,
                      warning = FALSE,
                      include = TRUE,
                      echo = TRUE,
                      error = TRUE,
                      fig.width = 11,
                      comment = NA)
```

```{r, echo=FALSE, include=FALSE}
library(MASS)
library(plyr) #this must be loaded before dplyr 
library(dplyr)
library(ggplot2)
library(broom)
library(knitr)
library(magrittr)
library(reshape2)
library(infotheo)
library(stats)
library(ggbiplot)
library(car)
library("Hiiragi2013")
set.seed(2013) #does this need to be set to 2013? why?
```

## (1) draft analysis plan for developing classifier for predating developmental stage (E3.25) based on gene expression

> 1. Check/remove rows with missing data  
  2. Randomly split data into training / test sets
  3. Check/Z-score ranges / distributions
  5. PCA training set [2]
  6. Train neural net on PC's
  7. Transform test set with stored means / sd's & PC rotations
  8. Test
  9. Rerun with cross validation if time allows

## (2) Write R script for performing EDA on the data set and then perform EDA
  
  a: raw intensity values from the original CEL files arranged in a matrix layout, where each column represents one hybridization, and rows stand for individual array features
  x: RMA normalized dataset in the assayData and annotation in the phenoData
  xq: single-cell gene expression levels measured by qPCR
  xql: single-cell gene expression measured by qPCR, with cells facing the blastocyst cavity labelled fluorescently
  
### Load Data
  
```{r}
data("x")
hw3A.features <- t(assayDataElement(x@assayData,'exprs'))
hw3A.classes <- x@phenoData@data$Embryonic.day
```

### Missing Data

```{r}
sapply(hw3A.features, function(x) sum(is.na(x)))
sapply(hw3A.classes, function(x) sum(is.na(x)))
```

> No missing data reported

### Split into Training & Test sets

```{r}
index <- sample(1:nrow(hw3A.features),round(0.8*nrow(hw3A.features)))
hw3A.features_train <- hw3A.features[index,]
hw3A.classes_train <- hw3A.classes[index]
hw3A.features_test <- hw3A.features[-index,]
hw3A.classes_test <- hw3A.classes[-index]
```

### Z-score training set

```{r}
train_sd = matrix()
train_mean = matrix()
hw3A.features_train_normalized <- hw3A.features_train
for(i in 1:ncol(hw3A.features_train))
  train_sd[i] = sd(hw3A.features_train[,i])
  train_mean[i] = mean(hw3A.features_train[,i])
  for(j in 1:nrow(hw3A.features_train))
    hw3A.features_train_normalized[j,i] = (hw3A.features_train[j,i] - train_mean[i]) / train_sd[i]
```

### Principal Component Analysis

```{r}
class <- as.factor(hw3A.classes_train)
data_pc <- prcomp(hw3A.features_train_normalized,center=FALSE,scale. = FALSE,retx=TRUE)
print(data_pc)
summary(data_pc)
plot(data_pc,type="l") 
```

> Over 98% of the variance is explained by the first 12 pc's, over 99% by the first 40

```{r}
first_12_pcs <- data_pc$x[,1:12]
first_12_pc_rotations <- data_pc$rotation[,1:12]
first_40_pcs <- data_pc$x[,1:40]
first_40_pc_rotations <- data_pc$rotation[,1:40]
```

## (3) Create version 2 of analysis plan that includes EDA and any revisions (if needed) based on EDA 

## (4) Write R script for analyzing the data 

## For submission A, provide the 2 analysis plans and 2 R scripts. 

## References
> 
[1] http://www.r-bloggers.com/fitting-a-neural-network-in-r-neuralnet-package/  
[2] http://stats.stackexchange.com/questions/121522/how-to-transform-test-set-to-the-pca-space-of-the-training-set-if-the-features  
[3] https://support.bioconductor.org/p/40730/
[4] https://bioconductor.org/packages/release/data/experiment/manuals/Hiiragi2013/man/Hiiragi2013.pdf
[5] http://127.0.0.1:17064/library/Hiiragi2013/doc/Hiiragi2013.pdf
